org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/ClassFileCache.xtend

```diff
@@ -0,0 +1,44 @@
+/*******************************************************************************
+ * Copyright (c) 2018 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.java.resource
+
+import java.util.HashMap
+import java.util.Map
+import org.eclipse.jdt.internal.compiler.env.IBinaryType
+import org.eclipse.xtext.naming.QualifiedName
+import org.eclipse.xtext.util.internal.EmfAdaptable
+
+/**
+ * @author Christian Dietrich - Initial contribution and API
+ * @since 2.14
+ */
+@EmfAdaptable
+class ClassFileCache {
+	
+	//TODO: concurrency
+	//TODO: clear
+	//TODO: weak?
+	val Map<QualifiedName, IBinaryType> cache = new HashMap()
```

Comment:
Should indeed be a concurrentHashMap

Reply:
My feeling is also that we should use String as the key since it is easier to create the same string for nested types / types with $ in their names compared to equal QN

Reply:
The cache should than expose methods like putIfAbsent or getOrCreate(key, Supplier)

Reply:
so you dont like the usage of the active annotation?

Reply:
No, sorry, that was not what I meant. On the cache itself there is containsKey, put and so on and so forth. That should be a one-shot operation that we can pass through to the underlying (Concurrent)HashMap

Reply:
am not sure how expensive building closures is at thew caller site

Reply:
and where should the pass go? there is no member in concurrent hashmap that accepts a supplier

Reply:
Ok, I understand the point. Since the values are idempotent, it would be ok to compute the stuff twice for sure. But in general, a single call on the map would be preferable.

Reply:
I was thinking about `java.util.Map.computeIfAbsent(K, Function<? super K, ? extends V>)` - you are right, it's not a supplier.

Reply:
regarding string vs fqn (downstream api like getExoirtedObjects uses QN

Reply:
Since it's a lossless conversion from QN to string but not the other way round, I'd prefer the cache to use strings nevertheless.

Reply:
but wont that indicate to use the QN that has "more" information in order to keep it?

Reply:
Maybe I misunderstood the semantics here.
Assuming I do have a class file name like `com.acme.Left$Right.class`. This could be the file that stems from a java source `class Left$Right { }` or `class Left { class Right{} }`. What would be the QN that is used from getExportedObjects? Would it be `[com, acme, Left, Right]` or `[com, acme, Left$Right]`? What we seem to put into the index is the latter `[com, acme, Left$Right]` - see XbaseQualifiedNameProvider, but we do seem to disagree on this, e.g. see JvmIdentifiableQualifiedNameProvider. If we can unambigously recreate the right names from the file names, the QN would be the better choice.

Reply:
Question is: can we?

Reply:
i dont know. i assue that jdt uses Outer$Inner in one source segment

Reply:
=> we create the QualifiedNames fromwhat jdt asks us